CDAP 8516 Refactor UGIProvider to take NamespacedEntityId

rohitsinha54 opened the pull request:
- Refactor UGIProvider to take NamespacedEntityId
- Removed the call made to OwnerAdmin in DefaultImpersonator which can be at remote location.
- Please ignore the commented out UGIProviderTest for now. I am working on updating it to be in line with the changes. Opening the PR for review of changes
- Tested on cluster.

Issue: https://issues.cask.co/browse/CDAP-8516
Build: http://builds.cask.co/browse/CDAP-RUT597-1

Commit:
CDAP 8516 Refactor UGIProvider to take NamespacedEntityId

Address comments

rohitsinha54 commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -1,238 +1,247 @@
-/*
```
Please ignore for now. Working on updating this test to be in line with the changes.

anew commented on cdap-common/src/main/java/co/cask/cdap/common/kerberos/ImpersonationInfo.java:
```diff
@@ -29,7 +30,13 @@
+    if (Strings.isNullOrEmpty(principal)) {
+      throw new IllegalArgumentException("A principal must be provided");
+    }
+    if (Strings.isNullOrEmpty(principal)) {
```
change principal to keytabUri

anew commented on cdap-common/src/main/java/co/cask/cdap/common/kerberos/ImpersonationOpInfo.java:
```diff
@@ -0,0 +1,49 @@
+ * A wrapper which wraps around the {@link co.cask.cdap.proto.id.NamespacedEntityId} on which impersonation needs to
+ * be performed and the type of operation {@link ImpersonatedOpType} which will be performed.
+ */
+public class ImpersonationOpInfo {
```
I am getting a little confused between this class and the ImpersonationInfo. This class really represents an impersonation request, while the other one is a response to that, right? Perhaps you can rename this to ImpersonationRequest?

anew commented on cdap-security/src/main/java/co/cask/cdap/security/impersonation/DefaultImpersonator.java:
```diff
@@ -66,6 +57,7 @@ public DefaultImpersonator(CConfiguration cConf, UGIProvider ugiProvider,
   public <T> T doAs(NamespacedEntityId entityId, Callable<T> callable,
                     ImpersonatedOpType impersonatedOpType) throws Exception {
     UserGroupInformation ugi = getUGI(entityId, impersonatedOpType);
+    LOG.debug("Performing doAs with UGI {}", ugi);
```
is this log message useful if we can't tell what it it performs?

anew commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -109,91 +114,91 @@ public static void finish() {
-      httpService.stopAndWait();
-    }
-  }
+//  @Test
```
why are all tests commented out?

rohitsinha54 replied:
I am working on updating the unit test for the changes.

anew commented on cdap-security/src/main/java/co/cask/cdap/security/impersonation/DefaultUGIProvider.java:
```diff
@@ -63,8 +79,42 @@
+      // remote location as in RemoteUGIProvider. Looking up namespace meta there will make a trip to master followed
+      // by one to get the credentials. Hence do it here in the DefaultUGIProvider which is running in master.
+      // Although, this will cause cache miss for explore implementation if the namespace config doesn't allow
+      // impersonating the namespace owner for explore queries but that a trade-off to avoid multiple remote calls in
```
However, you are now making two lookups in the namespace store, right? First, to find out whether this namespace allows explore impersonation, then possibly later (line 110) again to retrieve the namespace's principal and keytab.

rohitsinha54 replied:
Yes, that's right. The owner lookup will make another query to namespace store.

anew replied:
Since the behavior is correct, just less efficient, can you add a todo and open a Jira to refactor it later?

rohitsinha54 replied:
I am not sure as what kind of refactoring we can do to improve this. One way I see is it pass the NamespaceConfig to the OwnerAdmin from the UGIProvider itself which to me looks bad as OwnerAdmin APIs should not be working with NamespaceConfig as such. Don't you agree ? Can you give some idea on what can be done to put in jira details ?

anew commented on cdap-security/src/main/java/co/cask/cdap/security/impersonation/DefaultUGIProvider.java:
```diff
@@ -63,8 +79,42 @@
+      try {
+        NamespaceConfig nsConfig = namespaceQueryAdmin.get(impersonationOpInfo.getEntityId().getNamespaceId())
+          .getConfig();
+        if (impersonationOpInfo.getImpersonatedOpType().equals(ImpersonatedOpType.EXPLORE) &&
```
perhaps you can check whether this is an explore query before retrieving the namespace meta? That saves you a read if it is not for explore.

rohitsinha54 replied:
Good point. 👍

anew reviewed (commented):
Done a pass. Please see my comments.

rohitsinha54 commented:
@anew : Addressed comments. Please have another look. Also, please ignore UGIProviderTest class as its commented out now. I am working in updating it. We can review it in an independent commit at the end before merge. Thanks.

anew commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -78,42 +89,46 @@
-  private static OwnerStore ownerStore = new MockOwnerStore();
-  private static OwnerAdmin ownerAdmin = new DefaultOwnerAdmin(cConf, ownerStore, namespaceClient);
-  private static KerberosPrincipalId hdfsKerberosPrincipalId;
+  //  private static File keytabFile;
```
remove this line?

anew commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -140,97 +155,205 @@ public static void finish() {
+    try {
+      provider.getConfiguredUGI(bobImpRequest);
+      Assert.fail("Expected IllegalArgumentException when getting UGI for " + bobUGIWithPrincipal);
+    } catch (IllegalArgumentException e) {
```
Is this the expected behavior? If the user does not exist (or the keytab), we throw IllegalArgument? Shouldn't it be a NotFoundException or something like that? I would expect that IllegalArgument indicates that the bob request is ill-formed. Such as invalid characters in entity name. Can you file a Jira and add a TODO for an improvement?

anew replied:
Does the remote UGI provider throw the same exception? They should behave the same way from a client's perspective.

anew commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -140,97 +155,205 @@ public static void finish() {
-//  }
+  @Test
+  public void testDefaultUGIProviderWithHDFSFiles() throws Exception {
+    Location hdfsKeytabDir = locationFactory.create("keytabs");
```
looks like there is a lot of duplication between this and the previous test case... Can you consolidate them? Fine to do it in a new PR if you want to get this merged.

rohitsinha54 replied:
Let me try to do it in this PR itself. If we punt on it I guess we will never get to it later.

anew commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -140,97 +155,205 @@ public static void finish() {
+
+  @Test
+  public void testRemoteUGIProvider() throws Exception {
+    // Starts a mock server to handle remote UGI requests
```
and similarly, this test could be equal to the previous two, except for setting up the ugi provider.

anew commented on cdap-security/src/test/java/co/cask/cdap/security/impersonation/UGIProviderTest.java:
```diff
@@ -140,97 +155,205 @@ public static void finish() {
+  public void testRemoteUGIProvider() throws Exception {
+    // Starts a mock server to handle remote UGI requests
+    final NettyHttpService httpService = NettyHttpService.builder("remoteUGITest")
+      .addHttpHandlers(Collections.singleton(new UGIProviderTestHandler()))
```
why does this use a different handler? Shouldn't it test actual remote UGI provider?

rohitsinha54 replied:
It is testing the RemoteUGI provider. If you see the UGI provider which is created is Remote. Only the handler is mock and not the actual ImpersonationHandler. Its not testing the ImpersonationHandelr logic. I am not completely sure as why it doesn't use the actual ImpersonationHandler. I think the idea is not to start app-fabric here and test the just the RemoteUGIProvider.

Nishith commented:
Looked at the last commit and that looks good to me. Let's merge this so we can start testing and if there are any more comments then we can address them in a different PR.

anew commented:
LGTM

rohitsinha54 commented:
Waiting for the build here to pass : http://builds.cask.co/browse/CDAP-RUT597-10
